### PR TITLE
Docs sidebar

### PR DESCRIPTION
**Pull request by nendow02**

### CODE REVIEW

File: _includes/docs-sidebar.html
Comment (krashanoff) [inline _includes/docs-sidebar.html:7]: You could pull this information directly from `{{ site.posts }}`.

File: _sass/_docs.scss
Comment (krashanoff) [inline _sass/_docs.scss:17]: Looking at how you styled the sidebar, this would explain the viewport clipping bug I mention at the start of this thread. Instead of using `overflow-x: hidden`, `padding: npx mpx;`, and properties of that nature, what you tried using [flexbox](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Flexible_Box_Layout/Basic_Concepts_of_Flexbox)?

Also, rather than `height: 100vh`, we could try using `height: 100%`.

File: _sass/_docs.scss
Comment (krashanoff) [inline _sass/_docs.scss:23]: To reiterate my previous comment, many of the styles specified here could be abstracted away with a [flex container](https://developer.mozilla.org/en-US/docs/Web/CSS/CSS_Flexible_Box_Layout/Basic_Concepts_of_Flexbox).

File: _docs/dev-setup.md
Comment (krashanoff) [inline _docs/dev-setup.md:13]: I think having the table of contents might still be important, but I'll leave this decision to you. If you want a quicker way of generating the table of contents, you can do so with `{:toc}` ([see here for more](https://kramdown.gettalong.org/converter/html.html#toc)).

File: _includes/docs-sidebar.html
Comment (krashanoff) [inline _includes/docs-sidebar.html:19]: I'm pretty sure a lot of the text manipulation you're doing here can be supplanted by `{% assign spacelessDown = item | slugify %}` ([read more here](https://jekyllrb.com/docs/liquid/filters/)).

File: _layouts/docs-default.html
Comment (krashanoff) [inline _layouts/docs-default.html:22]: I'm pretty sure that rather than manually check this, we could instead just specify the layout of the docs home ([`_docs/index.html`](https://github.com/uclaacm/teach-la-website/blob/docs_sidebar/_docs/index.html)) so that it doesn't inherit the default value we give it in our `_config.yml`. Saves the coder a little bit of time reading over the layout.

File: _docs/dev-setup.md
Comment (krashanoff) [inline _docs/dev-setup.md:3]: I wish there was a better way of doing this, but from a glance at the Jekyll docs, it doesn't appear that there is an easy way to get all the headers from a file without writing a script to do so. Any ideas?

**Review by krashanoff — CHANGES_REQUESTED** [review]
Nathan, thanks for all your work and patience getting this feature rolled out! There are a few significant details that'll need to be addressed before this is good to merge, and a handful of small implementation-level details that you should consider that I left in a thread under this review.

## Viewport Clipping

Right now the biggest issue (and honestly the only blocking one) is that the sidebar will clip out on certain viewports. For example, opening up developer tools here, we can see that the sidebar text has some strange clipping going on:

image

## Design Details

Last thing is that it might be a good idea to remove the off-grey background for the sidebar. At present it creates an unbalanced look on each docs post. I think removing the background would help out.

***

If you have any questions at all on any of my review here, don't hesitate to DM me on Slack. If you'd like to get together sometime outside of club hours, as well, I could help walk through some of the trickier changes.

Thanks again Nathan!

File: _docs/dev-setup.md
Comment (malsf21) [reply _docs/dev-setup.md:3]: Think you'd have to delve deeper into Ruby/Jekyll/Nokogiri? https://www.honeybadger.io/blog/automatically-generating-subnavigation-from-headings-in-jekyll/

File: _includes/docs-sidebar.html
Comment (malsf21) [reply _includes/docs-sidebar.html:7]: I think `{{site.docs}}` instead - the `posts` collection is something else!

**Comment by nendow02** [comment]
Okay I made the changes you guys requested. Let me know if anything needs to be fixed!

**Review by krashanoff — CHANGES_REQUESTED** [review]
Hey Nathan! Great work, especially on the custom plugin. Flexbox container seems to be handling what we might throw at it. For the rest of the changes, I have just a few minor edits:

## Redirects 404

Click on a link in the sidebar and it will 404. This has to do with the way the `docs-sidebar` include is implemented. I left a suggestion for this that *should* do what you want. I'd test locally before doing anything, though.

## Viewport Overflow

The viewport overflows right now on screens smaller than `1000px` but before reaching the mobile breakpoint. This is because of the way the width of the content container is set. I left a suggestion for this, too.

Other than that, looks good to me! Great work.

**Comment by nendow02** [comment]
Made the fixes, I had to go a different way on fixing the overflow because the alignment got weird but I think I fixed it by changing the white-space property, definitely let me know what you think though

**Comment by nendow02** [comment]
Never mind I figured out how to implement the solution you recommended, sorry about that.

File: _sass/_docs.scss
Comment (krashanoff) [inline _sass/_docs.scss:72]:
```suggestion
    max-width: 100vw - $sidebar-width;
    @media (max-width: $desktop-width) {
        max-width: 100vw - 3em;
    }
```

The beauty of SASS is that you can simply do math before preprocessing :+1:

**Review by krashanoff — APPROVED** [review]
Thanks for implementing the fixes I requested. Things look pretty good to merge imo, but if you want to go above and beyond, there's two more things you can do before clicking the merge button:

Basically if you click to toggle a category, weird things happen. You can see the below GIF for more.

animated demo

Should be a pretty quick trick in the styles for the sidebar.

That's pretty much the only fix I can think of! Good work Nathan, seriously appreciate it.

**Comment by nendow02** [comment]
Okay I made the changes, there does seem to be a small merge conflict however in the sass variables doc, I’m not quite sure how to fix that.

**Comment by krashanoff** [comment]
I handled the merge conflict for you. There are a few small details that I think would make the sidebar better in terms of QoL, but we'd have to save those for when we have more docs. For now, this is a good start to the sidebar, and I don't want to keep this PR hung up more than it needs to be. Great work Nathan! You are clear to merge.

Make sure to "squash and merge" when doing so.

**Pull request merged**
